suite/tests/client-interface/drx_sleep_scale-test.cpp

```diff
@@ -0,0 +1,261 @@
+#else
+            res = syscall(SYS_clock_nanosleep_time64, CLOCK_REALTIME, 0, &sleeptime64,
+                          &remaining64);
+#endif
+        } else {
+            res = syscall(SYS_nanosleep, &sleeptime, &remaining);
+        }
+        if (res != 0) {
+            assert(errno == EINTR);
+            // Ensure the remaining time was deflated.
+#ifndef X64
+            if (clock_version)
+                assert(remaining64.tv_sec <= sleeptime64.tv_sec);
+            else
+#endif
+                assert(remaining.tv_sec <= sleeptime.tv_sec);
+            ++eintr_count;
+        }
+    }
+    assert(eintr_count > 0);
+    std::cerr << "eintrs=" << eintr_count << "\n";
+    return reinterpret_cast<void *>(sleep_count);
+}
+
+static int64_t
+do_some_work(bool clock_version)
+{
+    pthread_t thread;
+    void *retval;
+    pthread_mutex_init(&lock, NULL);
+    pthread_cond_init(&condvar, NULL);
+
+    pthread_mutex_lock(&lock);
+    child_ready = false;
+    pthread_mutex_unlock(&lock);
+    child_should_exit.store(false, std::memory_order_release);
+
+    int res = pthread_create(&thread, NULL, thread_routine,
+                             reinterpret_cast<void *>(clock_version));
+    assert(res == 0);
+    // Wait for the child to start running.
+    pthread_mutex_lock(&lock);
+    while (!child_ready)
+        pthread_cond_wait(&condvar, &lock);
+    pthread_mutex_unlock(&lock);
+    // Now take some time doing work so we can measure how many sleeps the child
+    // accomplishes in this time period.
+    constexpr int ITERS = 10000000;
+    double val = static_cast<double>(ITERS);
+    for (int i = 0; i < ITERS; ++i) {
+        val += sin(val);
+        // Test interrupting the thread's sleeps.
+        if (i % (ITERS / 20) == 0 ||
+            // Add a use of val so it's not optimized away.
+            val == 0.) {
+            pthread_kill(thread, SIGUSR1);
+        }
+    }
+    // Clean up.
+    child_should_exit.store(true, std::memory_order_release);
+    res = pthread_join(thread, &retval);
+    assert(res == 0);
+    pthread_cond_destroy(&condvar);
+    pthread_mutex_destroy(&lock);
+    return reinterpret_cast<int64_t>(retval);
+}
+
+static void
+event_exit(void)
+{
+    bool ok = drx_unregister_time_scaling();
+    assert(ok);
+    drx_exit();
+    dr_fprintf(STDERR, "client done\n");
+}
+
+static int64_t
+test_sleep(bool clock_version, int scale)
+{
+    std::string dr_ops("-stderr_mask 0xc -client_lib ';;" + std::to_string(scale) + "'");
+    if (!my_setenv("DYNAMORIO_OPTIONS", dr_ops.c_str()))
+        std::cerr << "failed to set env var!\n";
+    dr_app_setup_and_start();
+    int64_t count = do_some_work(clock_version);
+    dr_app_stop_and_cleanup();
+    return count;
+}
+
+static void
+test_sleep_scale()
+{
+    int64_t sleeps_default = test_sleep(/*clock_version=*/true, 1);
+    constexpr int SCALE = 100;
+    int64_t sleeps_scaled = test_sleep(/*clock_version=*/true, SCALE);
```

Comment:
If an app checks unix timestamps before and after the sleep, it would look like a faulty alarm clock to the app right? Can't do anything about that...